### PR TITLE
Model submission for CERBERUS ANYmal sensor configuration 1 from RSL-ETH Zurich

### PR DESCRIPTION
**Pull request by nkoenig**
his Pull Request submits the CERBERUS ANYmal Sensor Configuration 1 robot.

The following packages are added:

* **cerberus\\_anymal\\_sensor\\_config\\_1:** this package contains the robot’s sdf model and the needed launch files for the simulation. The package moreover contains a Readme file as well as the requested [specification.md](http://specification.md) file, which contains the instructions needed to use this vehicle\
* **cerberus\\_anymal\\_ignition\\_gazebo\\_plugin:** this package contains an ignition gazebo plugin needed to simulate and control the 12 actuators that enable the robot to move. This plugin is released as `.so` file.

Please note that, as stated in `cerberus_anymal_sensor_config_1/specifications.md`, another repository must be used to control the robot, given an input twist to follow. This repository is publicly available on the GitHub page of our lab: [cerberus\\_anymal\\_locomotion](https://github.com/leggedrobotics/cerberus_anymal_locomotion) and it contains all the needed instructions.

The last point worth to mention here is the decision to temporary comment out the “battery” plugin in the file `cerberus_anymal_sensor_config_1/launch/spawner.rb`. This is due to an unexpected behavior of this plugin, which freezes all the joints and does not enable to command a desired torque. We opened and [issue](https://bitbucket.org/ignitionrobotics/ign-gazebo/issues/55/battery-plugin-freezes-joints) about this problem and will enable the plugin again when the it will be fixed.

### CODE REVIEW

**Comment by Samu24** [comment]
@nkoenig We could reproduce the issue heuristicus mentioned. We are looking into this and will provide a fix.

**Comment by Samu24** [comment]
Open PR #417

**Comment by nkoenig** [comment]
The [battery joint issue](https://github.com/ignitionrobotics/ign-gazebo/issues/55) has been resolved [here](https://github.com/ignitionrobotics/ign-gazebo/pull/120).

**Comment by heuristicus** [comment]
I've been trying out this model and have noticed something strange with the `downward_realsense`. The top half of the image is cut off, both in the raw image and in the depth image.

This is as viewed from rviz.
Screenshot from 2020-05-27 13-04-59

A sample of the output received on the topic is:

```
rostopic echo -n1 /anymal/downward_realsense/depth/image_raw
header: 
  seq: 11130
  stamp: 
    secs: 1123
    nsecs: 200000000
  frame_id: "anymal/base/downward_realsense"
height: 240
width: 320
encoding: "32FC1"
is_bigendian: 0
step: 1280
data: [0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255, 0, 0, 128, 255...
```

**Comment by acschang** [comment]
The initial assessment of the CERBERUS ANYmal model is complete, however there are some issues to resolve before this model can be merged and utilized in a competitive setting:

- Please add a gas sensor to this model for consistency with other robot models in the SubT Tech Repo, even if the real robot does not include a gas sensor.
- The camera resolution and intrinsic parameters do not seem to match the FLIR Blackfly S datasheet. Please match the parameters.
- The BatteryPlugin issue should be fixed in this PR. Please add the BatteryPlugin to model the robot's battery life of 4200 seconds.
- Does the `robot_base_twist` topic provide ground-truth velocity? Ground-truth data cannot be available as ROS topics to the competitors' solutions. The `robot_base_twist` topic should either be degraded to represent filtered IMU data (like the odometry on other UGVs), or the controller should be included in the model, using Ignition topics rather than ROS topics. Please let us know if you'd like to discuss in more depth.

Please address the issues noted above and then submit a pull request to update your 
submission by **5 JUNE**. The submissions are expected to be incorporated into the SubT Virtual Testbed pending a successful review."

**Comment by Samu24** [comment]
Hi @acschang 

The first three points are straightforward for us to address. We will improve them and provide a PR.

For the forth point:
Yes the `robot_base_twist` topic is providing the twist read from the simulation i.e. ground truth linear and angular velocity. We need this information in our controller. However, we cannot put together the controller code with igntion specific code because of protobuf version mismatch which results in compiler errors. This means we can also not use the ignition transport system. A possible solution would be to add e.g. gaussian noise to the published ROS topic.

Let use know what you think or if you can think of a better solution for the problem.

**Comment by acschang** [comment]
Please try applying the same noise model used for the [UAV Velocity Controller](https://github.com/ignitionrobotics/ign-gazebo/blob/de5b025968c9cf0cfbfd8a852458482e87c70c6c/src/systems/multicopter_control/MulticopterVelocityControl.cc) to the `robot_base_twist` data and use the noise parameters defined here. If that causes issues with your controller, we can re-evaluate.

**Comment by heuristicus** [comment]
An update on the downward realsense issue, it appears to be caused by the `base_fixed_joint_lump__downward_realsense_camera_visual_3` visual component of the model. Removing the visual gives the full image as expected.

With some trial and error I was able to fix the issue by modifying `model.sdf` as in this patch, which changes the x coordinate of the realsense visual from 0.388757 to 0.387757.

```
diff --git a/submitted_models/cerberus_anymal_b_sensor_config_1/model.sdf b/submitted_models/cerberus_anymal_b_sensor_config_1/model.sdf
index 86d3baa6..4f4509d2 100644
--- a/submitted_models/cerberus_anymal_b_sensor_config_1/model.sdf
+++ b/submitted_models/cerberus_anymal_b_sensor_config_1/model.sdf
@@ -155,7 +155,7 @@
         </material>
       </visual>
       <visual name="base_fixed_joint_lump__downward_realsense_camera_visual_3">
-        <pose>0.388757 -0.000224 0.057675 -0.009501 0.382227 0</pose>
+        <pose>0.387757 -0.000224 0.057675 -0.009501 0.382227 0</pose>
         <geometry>
           <mesh>
             <scale>0.001 0.001 0.001</scale>
```

**Comment by Samu24** [comment]
@acschang Open PR #437

**Comment by Samu24** [comment]
@heuristicus I think the problem is that the mesh is interfering part of sensor. We had this issue with the front cam as well. I will move the realsense slightly as you propose. This should give a clear image.

**Comment by heuristicus** [comment]
I have found a further issue which relates to the realsense. Displaying the `/anymal/downward_realsense/points` topic in rviz displays it in the expected orientation. However, using the depth image and colour image to construct the pointcloud (with rviz DepthCloud display)  and display it, the cloud is displayed incorrectly. I believe this is because of the orientation of the `anymal/base/downward_realsense` frame in which the raw and depth images are published.

In this image, you can see both displays at once, and one of them is clearly wrong.

Screenshot from 2020-06-03 14-58-30

Here is a close up on the `downward_realsense` frame:

Screenshot from 2020-06-03 15-00-47

This indicates to me that the orientation of the frame is not correct, or that there needs to be a separate frame in which the depth image is published. As I understand, the z axis of the depth image frame should be pointing directly out from the camera, so that the z values represent the depth of the points.

**Comment by heuristicus** [comment]
A fix for the above sensor issue is to add the following to `vehicle_topics.launch`:

```
    <node
      pkg="subt_ros"
      type="optical_frame_publisher"
      name="optical_frame_publisher">
      <remap from="input/image" to="downward_realsense/image_raw" />
      <remap from="output/image" to="downward_realsense/optical/image_raw" />
      <remap from="input/camera_info" to="downward_realsense/camera_info" />
      <remap from="output/camera_info" to="downward_realsense/optical/camera_info" />
    </node>

    <node
      pkg="subt_ros"
      type="optical_frame_publisher"
      name="optical_frame_publisher_depth"
      args="depth">
      <remap from="input/image" to="downward_realsense/depth/image_raw" />
      <remap from="output/image" to="downward_realsense/optical/depth" />
    </node>
```

File: submitted_models/cerberus_anymal_b_sensor_config_1/launch/spawner.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
+def spawner(_name, _modelURI, _worldName, _x, _y, _z, _roll, _pitch, _yaw)
+  <<-HEREDOC
+  <plugin name="ignition::launch::GazeboFactory"
```
Comment (azeey) [inline submitted_models/cerberus_anymal_b_sensor_config_1/launch/spawner.rb:3]: Please merge with `master` and change `<plugin>` to `<spawn name='#{_name}'>`

File: submitted_models/cerberus_anymal_b_ignition_gazebo_plugin_1/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(cerberus_anymal_b_ignition_gazebo_plugin_1)
+
+find_package(catkin REQUIRED)
+
+catkin_package()
+
+install(FILES lib/libAnymalBControlPlugin1.so
```
[the remark this answers is not in the document]
Comment (azeey) [reply submitted_models/cerberus_anymal_b_ignition_gazebo_plugin_1/CMakeLists.txt:8]: https://github.com/osrf/subt/pull/383/commits/00c3df9f7b20bd60e63e4d91e43efb4daca2fa56

File: submitted_models/cerberus_anymal_b_sensor_config_1/launch/spawner.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+      <!-- Publish robot state information -->
+      <plugin filename="libignition-gazebo-pose-publisher-system.so"
+        name="ignition::gazebo::systems::PosePublisher">
+        <publish_link_pose>true</publish_link_pose>
+        <publish_sensor_pose>true</publish_sensor_pose>
+        <publish_collision_pose>false</publish_collision_pose>
+        <publish_visual_pose>false</publish_visual_pose>
+        <publish_nested_model_pose>#{$enableGroundTruth}</publish_nested_model_pose>
+      </plugin>
```
[the remark this answers is not in the document]
Comment (azeey) [reply submitted_models/cerberus_anymal_b_sensor_config_1/launch/spawner.rb:52]: https://github.com/osrf/subt/pull/383/commits/adc0b90c2eda2cfb72841a60d04a5d427ed3c973

File: submitted_models/cerberus_anymal_b_sensor_config_1/launch/vehicle_topics.launch [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  <arg name="name"/>
+  <arg name="link_name"       value="base"/>
+  <arg name="sensor_prefix"   value="/world/$(arg world_name)/model/$(arg name)/link/$(arg link_name)/sensor"/>
+  <group ns="$(arg name)">
+    <!-- Robot's state -->
+    <node
+      pkg="ros_ign_bridge"
+      type="parameter_bridge"
+      name="ros_ign_bridge_pose"
+      args="/model/$(arg name)/pose@geometry_msgs/TransformStamped[ignition.msgs.Pose">
```
[the remark this answers is not in the document]
Comment (azeey) [reply submitted_models/cerberus_anymal_b_sensor_config_1/launch/vehicle_topics.launch]: https://github.com/osrf/subt/pull/383/commits/adc0b90c2eda2cfb72841a60d04a5d427ed3c973

File: submitted_models/cerberus_anymal_b_sensor_config_1/launch/vehicle_topics.launch [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      <remap from="/model/$(arg name)/pose" to="pose"/>
+    </node>
+
```
[the remark this answers is not in the document]
Comment (azeey) [reply submitted_models/cerberus_anymal_b_sensor_config_1/launch/vehicle_topics.launch:16]: https://github.com/osrf/subt/pull/383/commits/adc0b90c2eda2cfb72841a60d04a5d427ed3c973

File: submitted_models/cerberus_anymal_b_sensor_config_1/launch/spawner.rb (hunk 1 above)
Comment (azeey) [reply submitted_models/cerberus_anymal_b_sensor_config_1/launch/spawner.rb]: https://github.com/osrf/subt/pull/383/commits/dd51902c199ecd0b7e388df7ce40d1b4e59a7709

**Comment by azeey** [comment]
Please note that I have changed to depth image topic from `downward_realsense/depth/image_raw` to `downward_realsense/depth` in https://github.com/osrf/subt/pull/383/commits/3f5eb02d91d516357d035046a6b6fdf4a2cb79a8. This makes DepthCloud in Rviz work without any topic relays.

**Comment by nkoenig** [comment]
I'm merging this PR into the cave_feature_release3 branch. This robot will be available when the feature release branch is merged into master.

**Pull request merged**